Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift

```diff
@@ -0,0 +1,473 @@
+                identifier: "org.swift.docc.DuplicateParameterDocumentation",
+                summary: "Parameter \(parameter.name.singleQuoted) is already documented",
+                notes: notes
+            ),
+            possibleSolutions: [
+                Solution(
+                    summary: "Remove duplicate parameter documentation",
+                    replacements: parameter.range.map { [Replacement(range: adjusted($0), replacement: "")] } ?? []
+                )
+            ]
+        )
+    }
+    
+    private func missingParameterProblem(name: String, before nextParameter: Parameter?, standalone: Bool, location: SourceLocation?, symbolKind: SymbolGraph.Symbol.Kind?) -> Problem {
+        let solutions: [Solution]
+        if let insertLocation = nextParameter?.range?.lowerBound ?? location {
+            let extraWhitespace = "\n" + String(repeating: " ", count: (nextParameter?.range?.lowerBound.column ?? 1 + (standalone ? 0 : 2) /* indent items in a parameter outline by 2 spaces */) - 1)
+            let replacement: String
+            if nextParameter != nil {
+                // /// - Paramaters:
+                // ///   - nextParameter: Description
+                //      ^inserting "- parameterName: placeholder\n  "
+                //                                              ^^^^ add newline after to insert before the other parameter
+                replacement = Self.newParameterDescription(name: name, standalone: standalone) + extraWhitespace
+            } else {
+                // /// - Paramaters:
+                // ///   - otherParameter: Description
+                //                                    ^inserting "\n  - parameterName: placeholder"
+                //                                                ^^^^ add newline before to insert after the last parameter
+                replacement = extraWhitespace + Self.newParameterDescription(name: name, standalone: standalone)
+            }
+            solutions = [
+                Solution(
+                    summary: "Document \(name.singleQuoted) parameter",
+                    replacements: [
+                        Replacement(range: adjusted(insertLocation ..< insertLocation), replacement: replacement)
+                    ]
+                )
+            ]
+        } else {
+            solutions = []
+        }
+        
+        return Problem(
+            diagnostic: Diagnostic(
+                source: location?.source ?? nextParameter?.range?.source,
+                severity: .warning,
+                range: adjusted(location.map { $0 ..< $0 }),
+                identifier: "org.swift.docc.MissingParameterDocumentation",
+                summary: "Parameter \(name.singleQuoted) is missing documentation"
+            ),
+            possibleSolutions: solutions
+        )
+    }
+    
+    // MARK: Generated content
+    
+    private static let objcErrorDescription: [Markup] = [
```

Comment:
Is there a better place to save this generated content? Having text in the compiler source code seems like a maintenance problem waiting to happen. We might decide to change our style guide or possible an editor might want to make some changes to these text strings and now we need a new DocC compiler version/build.

And if we ever localized documentation someday, these English strings will be hard to track down and translate.

Does DocC have any text files or string files where we could save these sorts of literal text strings?

Reply:
Oh `embedInCode` sounds interesting. Looking forward to seeing that in use.